### PR TITLE
New Recipe: Next Best Offer for Banking 

### PR DESCRIPTION
**Pull request by bbejeck**
### Description 

 Addresses issue https://github.com/confluentinc/ksqldb-recipes/issues/110

### New recipe checklist

- [x] Full code validated in Confluent Cloud
- [x] SQL style/syntax consistent to existing recipes
- [x] Validate presentation with `python3 -m mkdocs serve`
- [x] File structure and content follow [template](https://github.com/confluentinc/ksqldb-recipes/tree/main/template)
- [x] Added recipe to [navigation](https://github.com/confluentinc/ksqldb-recipes/blob/main/mkdocs.yml)
- [x] Spell check, e.g. using `aspell`

### CODE REVIEW

File: docs/predictive-analytics/next-best-offer/index.md [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+# Next Best Offer - Anticipating Your Customer's Needs
+
+Consumers today are faced with never ending marketing messages from a variety of sources.  Often these messages are generic and don't have any consideration for the individual needs of the consumer.  This one-size-fits-all approach leads to poor conversion rates.  A better approach is to tailor offerings that take into consideration the interests of the consumer based on previous purchases or behavior.   This recipe demonstrates how to take existing customer information and provide a "Next Best Offer" to encourage sales and retain customers.
+
+## Step-by-step
+
+### Setup your Environment
+
+Provision a Kafka cluster in [Confluent Cloud](https://www.confluent.io/confluent-cloud/tryfree/?utm_source=github&utm_medium=ksqldb_recipes&utm_campaign=inventory).
```
Comment (ybyzek) [inline docs/predictive-analytics/next-best-offer/index.md:15]:
```suggestion
Provision a Kafka cluster in [Confluent Cloud](https://www.confluent.io/confluent-cloud/tryfree/?utm_source=github&utm_medium=ksqldb_recipes&utm_campaign=next_best_effort).
```

File: docs/predictive-analytics/next-best-offer/process.sql [hunk 2]
```diff
@@ -0,0 +1,82 @@
+-- Setup initial stream/table from Kafka topics
+SET 'auto.offset.reset'='earliest';
+
+CREATE STREAM CUSTOMER_ACTIVITY_STREAM (
```
Comment (ybyzek) [inline docs/predictive-analytics/next-best-offer/process.sql:4]:
```suggestion
CREATE STREAM customer_activity_stream (
```

File: docs/predictive-analytics/next-best-offer/process.sql [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    CUSTOMER_ID INTEGER KEY,
+    ACTIVITY_TYPE STRING,
+    PROPENSITY_TO_BUY DOUBLE
+   ) WITH (
+    KAFKA_TOPIC = 'CUSTOMER_ACTIVITY_STREAM',
+    VALUE_FORMAT = 'JSON',
+    PARTITIONS = 6
+);
+
+CREATE TABLE CUSTOMERS (
```
Comment (ybyzek) [inline docs/predictive-analytics/next-best-offer/process.sql:16]:
```suggestion
CREATE TABLE customers (
```

File: docs/predictive-analytics/next-best-offer/process.sql [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    GENDER STRING,
+    INCOME INTEGER,
+    FICO INTEGER
+) WITH (
+    KAFKA_TOPIC = 'CUSTOMERS_TABLE',
+    VALUE_FORMAT = 'JSON',
+    PARTITIONS = 6
+);
+
+CREATE TABLE OFFERS (
```
Comment (ybyzek) [inline docs/predictive-analytics/next-best-offer/process.sql:30]:
```suggestion
CREATE TABLE offers (
```

File: docs/predictive-analytics/next-best-offer/process.sql [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    OFFER_NAME STRING,
+    OFFER_URL STRING
+) WITH (
+    KAFKA_TOPIC = 'OFFERS_STREAM',
+    VALUE_FORMAT = 'JSON',
+    PARTITIONS = 6
+);
+
+-- Application logic
+CREATE STREAM NEXT_BEST_OFFER
```
Comment (ybyzek) [inline docs/predictive-analytics/next-best-offer/process.sql:41]:
```suggestion
CREATE STREAM next_best_offer
```

File: docs/predictive-analytics/next-best-offer/process.sql [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    WHEN ct.INCOME > 100000 AND ct.FICO < 700 AND cask.PROPENSITY_TO_BUY < 0.9 THEN 1
+    WHEN ct.INCOME < 50000 AND cask.PROPENSITY_TO_BUY < 0.9 THEN 2
+    WHEN ct.INCOME >= 50000 AND ct.FICO >= 600 AND cask.PROPENSITY_TO_BUY < 0.9 THEN 3
+    WHEN ct.INCOME > 100000 AND ct.FICO >= 700 AND cask.PROPENSITY_TO_BUY < 0.9 THEN 4
+    ELSE 5
+END AS OFFER_ID 
+FROM CUSTOMER_ACTIVITY_STREAM cask
+INNER JOIN CUSTOMERS ct ON cask.CUSTOMER_ID = ct.CUSTOMER_ID
+
+CREATE STREAM NEXT_BEST_OFFER_LOOKUP
```
Comment (ybyzek) [inline docs/predictive-analytics/next-best-offer/process.sql:64]:
```suggestion
CREATE STREAM next_best_offer_lookup
```

File: docs/predictive-analytics/next-best-offer/manual.sql [hunk 7]
```diff
@@ -0,0 +1,24 @@
+-- Manually insert data if source connectors are not run
+INSERT INTO CUSTOMERS (customer_id, first_name, last_name, email, gender, income, fico) VALUES  (1,'Waylen','Tubble','wtubble0@hc360.com','Male',403646, 465);
```
Comment (ybyzek) [inline docs/predictive-analytics/next-best-offer/manual.sql:2]: For consistency with stylization of other recipes, these stream/table names should all be lowercase.

CUSTOMERS --> customers
CUSTOMER_ACTIVITY_STREAM --> customer_activity_stream

etc

File: docs/predictive-analytics/next-best-offer/source.json [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+  "kafka.api.secret"         : "<my-kafka-api-secret>",
+  "connection.host"          : "<database-host-address>",
+  "connection.user"          : "<database-username>",
+  "connection.password"      : "<database-password>",
+  "database"                 : "<database-name>",
+  "collection"               : "<database-collection-name>",
+  "poll.await.time.ms"       : "5000",
+  "poll.max.batch.size"      : "1000",
+  "copy.existing"            : "true",
+  "output.data.format"       : "JSON"
```
Comment (ybyzek) [inline docs/predictive-analytics/next-best-offer/source.json:14]:
```suggestion
  "output.data.format"       : "JSON",
```

File: docs/predictive-analytics/next-best-offer/source.sql [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+  'kafka.api.secret'         = '<my-kafka-api-secret>',
+  'connection.host'          = '<database-host-address>',
+  'connection.user'          = '<database-username>',
+  'connection.password'      = '<database-password>',
+  'database'                 = '<database-name>',
+  'collection'               = '<database-collection-name>',
+  'poll.await.time.ms'       = '5000',
+  'poll.max.batch.size'      = '1000',
+  'copy.existing'            = 'true',
+  'output.data.format'       = 'JSON'
```
Comment (ybyzek) [inline docs/predictive-analytics/next-best-offer/source.sql:15]:
```suggestion
  'output.data.format'       = 'JSON',
```

File: docs/predictive-analytics/next-best-offer/docker-compose.yml [hunk 10]
```diff
@@ -0,0 +1,60 @@
+---
```
Comment (ybyzek) [inline docs/predictive-analytics/next-best-offer/docker-compose.yml:1]: Is this file needed?

**Review by ybyzek — COMMENTED** [review]
@bbejeck thanks! I made a first pass-thru (no code validation)

File: docs/predictive-analytics/next-best-offer/docker-compose.yml (hunk 10 above)
Comment (bbejeck) [reply docs/predictive-analytics/next-best-offer/docker-compose.yml:1]: Not really; removing it

**Comment by bbejeck** [comment]
@ybyzek I've addressed all comments - it's ready for another pass/validation

File: docs/predictive-analytics/next-best-offer/index.md [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+    nbo.INCOME,
+    nbo.FICO,
+    ot.OFFER_NAME,
+    ot.OFFER_URL
+FROM next_best_offer nbo
+INNER JOIN OFFERS ot
+ON nbo.OFFER_ID = ot.OFFER_ID;
+```
+
+
```
Comment (ybyzek) [inline docs/predictive-analytics/next-best-offer/index.md:209]: @bbejeck can these extraneous lines be removed?

File: docs/predictive-analytics/next-best-offer/index.md [hunk 12]
```diff
@@ -0,0 +1,214 @@
+---
+seo:
+  title: Next Best Offer — Anticipating Your Customer's Needs
```
Comment (ybyzek) [inline docs/predictive-analytics/next-best-offer/index.md:3]:
```suggestion
  title: Next Best Offer for Banking
```

File: docs/predictive-analytics/next-best-offer/index.md [hunk 13]
```diff
@@ -0,0 +1,214 @@
+---
+seo:
+  title: Next Best Offer — Anticipating Your Customer's Needs 
+  description: This recipe demonstrates how to use ksqlDB to present relevant offers to banking customers.
+---
+
+# Next Best Offer — Anticipating Your Customer's Needs
```
Comment (ybyzek) [inline docs/predictive-analytics/next-best-offer/index.md:7]:
```suggestion
# Next Best Offer for Banking
```

File: docs/predictive-analytics/next-best-offer/index.md [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+
+```json
+--8<-- "docs/predictive-analytics/next-best-offer/source.json"
+```
+
+--8<-- "docs/shared/manual_insert.md"
+
+### Run the stream processing application
+
+This application will perform a series of joins between event streams and tables to calculate the next best offer for a banking consumer, based on their activity. This should yield higher customer activity and satisfaction.
```
Comment (ybyzek) [inline docs/predictive-analytics/next-best-offer/index.md:32]: Since `consumer` is overloaded with Kafka and since the ksqlDB commands say `customer`, can this recipe `:%s/consumer/customer/g`?

File: docs/predictive-analytics/next-best-offer/process.sql [hunk 15]
```diff
@@ -0,0 +1,82 @@
+-- Setup initial stream/table from Kafka topics
```
Comment (ybyzek) [inline docs/predictive-analytics/next-best-offer/process.sql:1]: This comment doesn't correspond to the command below.  Either move comment down or delete

File: docs/predictive-analytics/next-best-offer/manual.sql [hunk 16]
```diff
@@ -0,0 +1,24 @@
+-- Manually insert data if source connectors are not run
```
Comment (ybyzek) [inline docs/predictive-analytics/next-best-offer/manual.sql:1]: This is a good comment but I'd like to look into doing this programmatically for all recipes (none of which have a comment in this file).  Let's please delete for now, for consistency.

File: docs/predictive-analytics/next-best-offer/index.md [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+   ) WITH (
+    KAFKA_TOPIC = 'CUSTOMER_ACTIVITY_STREAM',
+    VALUE_FORMAT = 'JSON',
+    PARTITIONS = 6
+);
+```
+
+Note that the `CUSTOMER_ID` field is the key in the stream's key-value pairs. You'll see why this is important in the next section.
+
+In a production setting, you would populate the stream's underlying topic either using a `KafkaProducer` application or from an external system via a [managed connector on Confluent Cloud](https://docs.confluent.io/cloud/current/connectors/index.html). For the purposes of this example, you'll manually insert records into the stream, using [INSERT VALUES](https://docs.ksqldb.io/en/latest/developer-guide/ksqldb-reference/insert-values/#insert-values) statements:
```
Comment (ybyzek) [inline docs/predictive-analytics/next-best-offer/index.md:73]: Even "for the purposes of this example", the primary path should be to use the connector.  Can you change the wording to reflect that `INSERT INTO` is only if no connector can be run in the user's case?

File: docs/predictive-analytics/next-best-offer/index.md [hunk 18]
```diff
@@ -0,0 +1,214 @@
...
+ct.FICO,
+CASE  
+    WHEN ct.INCOME > 100000 AND ct.FICO < 700 AND cask.PROPENSITY_TO_BUY < 0.9 THEN 1
+    WHEN ct.INCOME < 50000 AND cask.PROPENSITY_TO_BUY < 0.9 THEN 2
+    WHEN ct.INCOME >= 50000 AND ct.FICO >= 600 AND cask.PROPENSITY_TO_BUY < 0.9 THEN 3
+    WHEN ct.INCOME > 100000 AND ct.FICO >= 700 AND cask.PROPENSITY_TO_BUY < 0.9 THEN 4
+    ELSE 5
+END AS OFFER_ID 
+FROM customer_activity_stream cask
+INNER JOIN customers ct ON cask.CUSTOMER_ID = ct.CUSTOMER_ID
```
Comment (ybyzek) [inline docs/predictive-analytics/next-best-offer/index.md:178]: This ksqlDB statement fails, missing `;`

File: docs/predictive-analytics/next-best-offer/process.sql [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+ct.FICO,
+CASE
+    WHEN ct.INCOME > 100000 AND ct.FICO < 700 AND cask.PROPENSITY_TO_BUY < 0.9 THEN 1
+    WHEN ct.INCOME < 50000 AND cask.PROPENSITY_TO_BUY < 0.9 THEN 2
+    WHEN ct.INCOME >= 50000 AND ct.FICO >= 600 AND cask.PROPENSITY_TO_BUY < 0.9 THEN 3
+    WHEN ct.INCOME > 100000 AND ct.FICO >= 700 AND cask.PROPENSITY_TO_BUY < 0.9 THEN 4
+    ELSE 5
+END AS OFFER_ID 
+FROM customer_activity_stream cask
+INNER JOIN customers ct ON cask.CUSTOMER_ID = ct.CUSTOMER_ID
```
Comment (ybyzek) [inline docs/predictive-analytics/next-best-offer/process.sql:62]: This ksqlDB statement fails, missing `;`

File: docs/predictive-analytics/next-best-offer/process.sql [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+    nbo.CUSTOMER_ID,
+    nbo.PROPENSITY_TO_BUY,
+    nbo.ACTIVITY_TYPE,
+    nbo.INCOME,
+    nbo.FICO,
+    ot.OFFER_NAME,
+    ot.OFFER_URL
+FROM next_best_offer nbo
+INNER JOIN offers ot
+ON nbo.OFFER_ID = ot.OFFER_ID;
```
Comment (ybyzek) [inline docs/predictive-analytics/next-best-offer/process.sql:82]: What should this yield with the given `INSERT INTO` statements?  I see one record output.  Is that right?

**Review by ybyzek — CHANGES_REQUESTED** [review]
@bbejeck Ran through the recipe more thoroughly, thanks for the PR!

File: docs/predictive-analytics/next-best-offer/source.json [hunk 21]
```diff
@@ -0,0 +1,16 @@
+{
+  "connector.class"          : "MongoDbAtlasSource",
+  "name"                     : "recipe-mongodb-fleet_description",
+  "kafka.api.key"            : "<my-kafka-api-key>",
+  "kafka.api.secret"         : "<my-kafka-api-secret>",
+  "connection.host"          : "<database-host-address>",
+  "connection.user"          : "<database-username>",
+  "connection.password"      : "<database-password>",
+  "database"                 : "<database-name>",
+  "collection"               : "<database-collection-name>",
```
Comment (ybyzek) [inline docs/predictive-analytics/next-best-offer/source.json:10]: @bbejeck I just looked into topic naming conventions for this connector (docs: https://docs.confluent.io/cloud/current/connectors/cc-mongo-db-source.html) and it seems the topic name generated will be of the form `<prefix>.<database-name>.<collection-name>`.  So as written today, if the user launches this recipe, the topic names won't match between what the connector generates and what the ksqlDB app calls the topics per the param `KAFKA_TOPIC`.  Can you reconcile this so the recipe _can_ work OOTB if the user is actually using this source connector?

File: docs/predictive-analytics/next-best-offer/index.md [hunk 22]
```diff
@@ -0,0 +1,214 @@
...
+
+# Next Best Offer — Anticipating Your Customer's Needs
+
+Consumers today are faced with never-ending marketing messages from a variety of sources. Often these messages are generic and don't have any consideration for the individual needs of the consumer. This one-size-fits-all approach leads to poor conversion rates. A better approach is to tailor offerings that take into consideration the interests of the consumer, based on previous purchases or behavior. This recipe demonstrates how to take existing customer information and provide a "Next Best Offer" to encourage sales and retain customers.
+
+## Step-by-step
+
+### Set up your environment
+
+Provision a Kafka cluster in [Confluent Cloud](https://www.confluent.io/confluent-cloud/tryfree/?utm_source=github&utm_medium=ksqldb_recipes&utm_campaign=next_best_effort).
```
Comment (ybyzek) [inline docs/predictive-analytics/next-best-offer/index.md:15]:
```suggestion
Provision a Kafka cluster in [Confluent Cloud](https://www.confluent.io/confluent-cloud/tryfree/?utm_source=github&utm_medium=ksqldb_recipes&utm_campaign=next_best_offer).
```

File: docs/predictive-analytics/next-best-offer/process.sql [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+
+-- Application logic
+CREATE STREAM next_best_offer
+WITH (
+    KAFKA_TOPIC = 'NEXT_BEST_OFFER',
+    VALUE_FORMAT = 'JSON',
+    PARTITIONS = 6
+) AS
+SELECT 
+cask.ACTIVITY_ID,
```
Comment (ybyzek) [inline docs/predictive-analytics/next-best-offer/process.sql:48]: Should each of these `cask.` lines be indented a few spaces?

File: mkdocs.yml [hunk 24]
```diff
@@ -21,6 +21,7 @@ nav:
   - Predictive Analytics:
     - Optimize fleet management: predictive-analytics/fleet_management/index.md
     - Optimize omni-channel inventory: predictive-analytics/inventory/index.md
+    - Next best offer for banking: predictive-analytics/next-best-offer/index.md
```
Comment (ybyzek) [inline mkdocs.yml:24]: Hide in nav until PMM provides copy & graphics

```suggestion
    #- Next best offer for banking: predictive-analytics/next-best-offer/index.md
```

**Review by bbejeck — COMMENTED** [review]
committing suggestions

File: docs/predictive-analytics/next-best-offer/process.sql (hunk 20 above)
Comment (bbejeck) [reply docs/predictive-analytics/next-best-offer/process.sql:81]: It should have yielded more than one record output.  I've changed the `INSERT INTO` statements slightly to give  better output variety
Screen Shot 2022-01-18 at 4 28 22 PM

File: docs/predictive-analytics/next-best-offer/source.json (hunk 21 above)
Comment (bbejeck) [reply docs/predictive-analytics/next-best-offer/source.json]: Updated to use the Postgres connector as I think that's a more realistic connector for this use case.

**Comment by bbejeck** [comment]
@ybyzek updated and tested for comments

File: docs/predictive-analytics/next-best-offer/index.md [hunk 25]
```diff
@@ -0,0 +1,208 @@
...
+
+### Creating an event stream
+
+To get started, you'll first need to create a stream that contains the customer activity:
+
+```sql
+CREATE STREAM customer_activity_stream (
+    ACTIVITY_ID INTEGER,
+    IP_ADDRESS STRING,
+    CUSTOMER_ID INTEGER KEY,
```
Comment (mjsax) [inline docs/predictive-analytics/next-best-offer/index.md:61]: Nit: should we move this to the first line?

File: docs/predictive-analytics/next-best-offer/index.md [hunk 26]
```diff
@@ -0,0 +1,208 @@
...
+);
+```
+
+Note that the `CUSTOMER_ID` field is the key in the stream's key-value pairs. You'll see why this is important in the next section.
+
+In a production setting you'll populate the stream's underlying topic either with `KafkaProducer` application or from an external system using a [managed connector on Confluent Cloud](https://docs.confluent.io/cloud/current/connectors/index.html).  In case you don't have a connector set up, you can manually insert records into the stream with [INSERT VALUES](https://docs.ksqldb.io/en/latest/developer-guide/ksqldb-reference/insert-values/#insert-values) statements:
+
+
+```sql
+INSERT INTO customer_activity_stream (activity_id, ip_address, customer_id, activity_type, propensity_to_buy) VALUES (1,'121.219.110.170',1,'branch_visit',0.4);
```
Comment (mjsax) [inline docs/predictive-analytics/next-best-offer/index.md:77]: If we change the column order above, we would need to update these statements, too.

File: docs/predictive-analytics/next-best-offer/index.md [hunk 27]
```diff
@@ -0,0 +1,208 @@
...
+
+### Determining the next best offer
+
+Now you'll create the event stream that calculates the next best offer for your customers based on their activity. 
+
+#### Calculating the offer
+
+To perform the next offer calculation, create a stream that performs a join between the `CUSTOMER_ACTIVITY_STREAM` and the `CUSTOMERS` table:
+
+```sql
```
Comment (mjsax) [inline docs/predictive-analytics/next-best-offer/index.md:157]: Above we create the input stream first and insert value into it, before we create the table and add data. This implies that the stream data will have lower timestamps compared to the table data and thus would join agains an "empty" table, as it will be processed first. -- Should we switch both above to ensure that the stream-table joins works as expected?

It also applies to the "offer" table, ie, the input STREAM should be created (well, populated with data would be sufficient) last?

File: docs/predictive-analytics/next-best-offer/index.md [hunk 28]
```diff
@@ -0,0 +1,208 @@
...
+```
+
+The `CASE` statement is the workhorse for the query. It provides the next offer for the customer based on information resulting from the join. Note that you're using an `INNER JOIN` here because if the customer id isn't found in the `CUSTOMERS` table, there's no calculation to make. You'll notice that the result of the `CASE` statement is a single integer—the code for the offer to be made—so there's one final step left.
+
+#### Final results
+
+For the last step, create a query which contains the final results by joining the `NEXT_BEST_OFFER` stream with the `OFFERS` table:
+
+```sql
+CREATE STREAM next_best_offer_lookup
```
Comment (mjsax) [inline docs/predictive-analytics/next-best-offer/index.md:189]: In this query we join on offer_id which implies a repartitioning of the input stream. We could avoid this cost, by adding a `PARTITION BY OFFER_ID` (well we would need to full case statement...) in the first query.

Not sure if it would confuse users... -- as an alternative, we could also merge both queries into one; frankly, it might not be less confusing as we would need to use the CASE statement in the second join condition, too...

File: docs/predictive-analytics/next-best-offer/manual.sql [hunk 29]
```diff
@@ -0,0 +1,23 @@
+INSERT INTO customers (customer_id, first_name, last_name, email, gender, income, fico) VALUES  (1,'Waylen','Tubble','wtubble0@hc360.com','Male',403646, 465);
+INSERT INTO customers (customer_id, first_name, last_name, email, gender, income, fico) VALUES  (2,'Joell','Wilshin','jwilshin1@yellowpages.com','Female',109825, 705);
+INSERT INTO customers (customer_id, first_name, last_name, email, gender, income, fico) VALUES  (3,'Ilaire','Latus','ilatus2@baidu.com','Male',407964, 750);
+
+
+INSERT INTO customer_activity_stream (activity_id, ip_address, customer_id, activity_type, propensity_to_buy) VALUES (1,'121.219.110.170',1,'branch_visit',0.4);
```
Comment (mjsax) [inline docs/predictive-analytics/next-best-offer/manual.sql:6]: If we change the column order above, we would need to update these statements, too.

Also, should we do the INSERT INTO for the STREAM last (as mentioned above)?

File: docs/predictive-analytics/next-best-offer/process.sql [hunk 30]
```diff
@@ -0,0 +1,81 @@
+SET 'auto.offset.reset'='earliest';
+
+CREATE STREAM customer_activity_stream (
+    ACTIVITY_ID INTEGER,
+    IP_ADDRESS STRING,
+    CUSTOMER_ID INTEGER KEY,
```
Comment (mjsax) [inline docs/predictive-analytics/next-best-offer/process.sql:6]: as above?

File: docs/predictive-analytics/next-best-offer/index.md (hunk 27 above)
Comment (bbejeck) [reply docs/predictive-analytics/next-best-offer/index.md:156]: Yeah I overlooked that aspect when building this - I've updated to create the tables first as well as insert the data

**Review by mjsax — APPROVED** [review]

File: docs/predictive-analytics/next-best-offer/index.md [hunk 31]
```diff
@@ -0,0 +1,207 @@
...
+In the event stream you'll use for this recipe, each activity entry contains only the id for the customer. This is expected, as it's a common practice to have [normalized](https://en.wikipedia.org/wiki/Database_normalization) event streams.  But when it's time to analyze the data, it's important to have additional customer information to provide context for any analysts reviewing the results. You'll also need a table to use for the calculated offer.
+
+#### Creating the customer table and inserting records
+
+Create the table for customer information:
+
+```sql
+CREATE TABLE customers (
+    CUSTOMER_ID INTEGER PRIMARY KEY,
+    FIRST_NAME STRING,
```
Comment (ybyzek) [inline docs/predictive-analytics/next-best-offer/index.md:65]: Per https://github.com/confluentinc/ksqldb-recipes/issues/146#issuecomment-1020484699, can this be updated to use `VARCHAR` instead of `STRING` (here and throughout the recipe)?

Comment (bbejeck) [reply docs/predictive-analytics/next-best-offer/index.md]: @ybyzek - I've updated to use `VARCHAR` vs. `STRING`

**Review by ybyzek — APPROVED** [review]
@bbejeck LGTM, thank you!
